pkg/loadbalancer/server.go

```diff
@@ -22,14 +22,17 @@ import (
 )
 
 type Server struct {
+	directConnectivity bool
 	tunnelManager *tunnelManager
 }
 
 var _ cloudprovider.LoadBalancer = &Server{}
 
-func NewServer() cloudprovider.LoadBalancer {
-	s := &Server{}
-	if runtime.GOOS == "darwin" || runtime.GOOS == "windows" || isWSL2() {
+func NewServer(directConnectivity bool) cloudprovider.LoadBalancer {
+	s := &Server{
+		directConnectivity: directConnectivity,
+	}
+	if directConnectivity || runtime.GOOS == "darwin" || runtime.GOOS == "windows" || isWSL2() {
```

Comment:
hmm, this is `directConnectivy && containers running inside VM` 

I don't know from the top of my head how to detect `containers running inside VM` 🤔

Reply:
@BenTheElder you have any idea on how to identify if the containers are running inside a VM as in Mac or Windows?

Reply:
> … how to identify if the containers are running inside a VM as in Mac or Windows?

I'm not sure that that is a _valid_ distinction. The system either runs within a container or it doesn't; whether the container runtime is hosted directly on the _host_ or within a VM running on the _host_ is not really its concern?

When the readme states:

https://github.com/kubernetes-sigs/cloud-provider-kind/blob/f26b241ce573c53e81dfca8070167b0288626c03/README.md?plain=1#L176-L182

I took that to mean that the expectation should be that any load balancer _ports_ would be mapped, by Docker, to ports on the host; hence my proposal to enable the `tunnelManager` for the `directConnectivity ` scenario

https://github.com/kubernetes-sigs/cloud-provider-kind/blob/5e03531c9674276a444e7270a1452bfc3c03a048/pkg/loadbalancer/server.go#L35-L37

I don't understand why anyone would not wish to run the "cloud-provider-kind" system within a container - because `--restart=always` can be used, thus giving the system the same scope/lifetime as any corresponding Kind clusters. So, it strikes me that, the pertinent distinction is whether `--network host` or `--network kind` is used when running the container, which respectively correspond to the value of `internal`/`directConnectivity`. If running on a Linux _host_ then `--network host` would be used, because that facilitates direct IP and port registration behaviour, whereas a non-Linux _host_ is required to use `--network kind`, thus limiting the functionality to the host port maps created by Docker.

Reply:
I'm just a pretty busy these days, I want this change to land but I want also to get this as much scoped as possible to avoid possible edge cases and to have sane defaults, relying on users to follow recommendations never works XD ... just please have some patience, I will make time next week to finish this

Reply:
No worries. I greatly appreciate your help and insight, not to mention the general implementation of "cloud-provider-kind".

There's no rush on my account - I have a working system now - and it will take me a while to sort the CLA.

If you wish to abandon this pull-request and re-implement it yourself, then that would be find with me - you did most of the work, I just did some typing.

Reply:
> @BenTheElder you have any idea on how to identify if the containers are running inside a VM as in Mac or Windows?

Doing this reliably from within the container is problematic, there are things we can sniff but nothing guaranteed.
Instead probably detecting the container itself from the docker / nerdctl API could work, but I don't think we can support this for podman.

> I'm not sure that that is a valid distinction. The system either runs within a container or it doesn't; whether the container runtime is hosted directly on the host or within a VM running on the host is not really its concern?

The networking and filesystem implications are substantially different, for starters. When using something like docker desktop there is an expectation to bridge the boundary between the host and the containers in the VM, we're a lot more limited in our options if we're running entirely within the VM.

It's true that we use docker portmap currently though.

> Indeed; but - unless people go to the trouble of defining a systemd unit, or other automated launch mechanism, or manually launch the "cloud-provider-kind" executable every time they want to use it - why wouldn't everyone want to ensure that the system was running whenever their Kind cluster[s] are running (i.e. whenever the Docker daemon is running, which may be trivially achieved by running it within a container using --restart=always)?

This isn't necessarily possible in podman and nerdctl, which we also support[^1].

We could provide a reference systemd unit, however currently the expectation is that you'll run this in the background while you develop.

[^1]: sort of, it's not that simple, but in broad strokes this is a rather docker specific thing last I checked, podman and nerdctl do not have the same restart functionality.